{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": 1,
   "id": "f6d2f9d8-b5cf-4f18-ba51-decb9420efef",
   "metadata": {},
   "outputs": [],
   "source": [
    "import dendropy, os, subprocess, shutil\n",
    "import pandas as pd\n",
    "import numpy as np\n",
    "import matplotlib.pyplot as plt"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "id": "91c42cde-8c47-448e-8736-c7102889f9ee",
   "metadata": {},
   "outputs": [],
   "source": [
    "subtree_inputs = pd.read_csv(\n",
    "    '240816to21_results.passfilter.subtree.csv', index_col=0)\n",
    "\n",
    "subtree_list = np.unique(\n",
    "    subtree_inputs.loc[:, 'subtree'])\n",
    "subtree_list = subtree_list[subtree_list != 'NONE']"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 14,
   "id": "aa1bb573-573d-4051-a763-43b79e26336d",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "55259"
      ]
     },
     "execution_count": 14,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "np.sum(subtree_inputs.subtree != 'NONE', axis=0)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "id": "0de4811d-456d-486b-b13a-6f3a8c546575",
   "metadata": {},
   "outputs": [],
   "source": [
    "background_strains = []\n",
    "np.random.seed(42)\n",
    "for subtree_name in subtree_list:\n",
    "    # randomly select 20 strains for background\n",
    "    strains = np.random.choice(\n",
    "        subtree_inputs.loc[subtree_inputs.subtree == subtree_name].index, size=20, replace=False)\n",
    "    background_strains.append(\n",
    "        pd.DataFrame(\n",
    "            data=[strains, [subtree_name for i in strains]],\n",
    "            index=['strain', 'subtree']).T)\n",
    "background_strains = pd.concat(background_strains, axis=0).reset_index()"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "fbdc7cb3-c99f-49b8-ae7e-3b85263af4a2",
   "metadata": {
    "jp-MarkdownHeadingCollapsed": true
   },
   "source": [
    "## Subtree generation"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "e540c911-1d25-444c-87e7-ab5062226c77",
   "metadata": {},
   "source": [
    "Generate fastas"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "id": "38cb71cf-b1b6-4b02-a031-238470515925",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Submitted batch job 46565495\n",
      "Submitted batch job 46565496\n",
      "Submitted batch job 46565497\n",
      "Submitted batch job 46565498\n",
      "Submitted batch job 46565509\n",
      "Submitted batch job 46565510\n",
      "Submitted batch job 46565511\n",
      "Submitted batch job 46565512\n",
      "Submitted batch job 46565513\n",
      "Submitted batch job 46565514\n",
      "Submitted batch job 46565515\n",
      "Submitted batch job 46565516\n",
      "Submitted batch job 46565518\n",
      "Submitted batch job 46565519\n"
     ]
    }
   ],
   "source": [
    "work_dir = 'scratch/supertree_construction/'\n",
    "os.makedirs(work_dir, exist_ok=True)\n",
    "\n",
    "for subtree_name in subtree_list:\n",
    "    subtree_strains = subtree_inputs.loc[subtree_inputs.subtree == subtree_name].index\n",
    "    # collect paths (+ canettii outgroup)\n",
    "    subtree_data = [[\n",
    "        'canettii',\n",
    "        '240805_qingyun_sra/SRR10522783/results/SRR10522783.breseq.vcf',\n",
    "        '240805_qingyun_sra/SRR10522783/results/SRR10522783.miss.breseq.zarr']]\n",
    "    # combine subtree strains with background strains\n",
    "    subtree_input = np.unique(np.concatenate([subtree_strains, background_strains.strain]))\n",
    "    for index in subtree_input:\n",
    "        row_data = [index,\n",
    "             f'{subtree_inputs.loc[index, \"batch\"]}/{index}/results/{index}.breseq.vcf',\n",
    "             f'{subtree_inputs.loc[index, \"batch\"]}/{index}/results/{index}.miss.breseq.zarr']\n",
    "        subtree_data.append(row_data)\n",
    "    subtree_df = pd.DataFrame(\n",
    "        data=subtree_data,\n",
    "        columns=['label', 'vcf_path', 'miss_path'])\n",
    "    # write df to working directory, but remove anything that's already there\n",
    "    if os.path.exists(f'{work_dir}/{subtree_name}'):\n",
    "        shutil.rmtree(f'{work_dir}/{subtree_name}')\n",
    "    os.makedirs(f'{work_dir}/{subtree_name}', exist_ok=True)\n",
    "    subtree_df.to_csv(f'{work_dir}/{subtree_name}/subtree_inputs.csv')\n",
    "    # run fasta generation command for subtree\n",
    "    cmd = f'\\\n",
    "        sbatch -c 40 -p sapphire -t 2:00:00 --mem=80G -o {work_dir}/{subtree_name}/fasta-%j.out --wrap=\"\\\n",
    "        module load Mambaforge/22.11.1-fasrc01 && conda activate mtb_isolates && \\\n",
    "        export PATH=\\\"/n/boslfs02/LABS/sfortune_lab/Lab/culviner/bin/mtbvartools/scripts:/n/home12/pculviner/.conda/envs/mtb_isolates/bin/:$PATH\\\" && \\\n",
    "        export PYTHONPATH=\\\"${{PYTHONPATH}}:/n/boslfs02/LABS/sfortune_lab/Lab/culviner/bin/mtbvartools\\\" && \\\n",
    "        write_snp_fastas.py \\\n",
    "        --input-csv {work_dir}/{subtree_name}/subtree_inputs.csv \\\n",
    "        --input-fasta /n/boslfs02/LABS/sfortune_lab/Lab/culviner/notebooks/240807_isolate_tree_expansion/pipeline_inputs/Mtb_h37rv.fasta \\\n",
    "        --inputs-dir /n/boslfs02/LABS/sfortune_lab/Lab/culviner/notebooks/240807_isolate_tree_expansion/scratch/ \\\n",
    "        --out-dir {work_dir}/{subtree_name}/ \\\n",
    "        --output {subtree_name} \\\n",
    "        --mask rlc_plus_lowmap_marin.bed \\\n",
    "        --miss-threshold 0.5 \\\n",
    "        --local-threads 40\"'\n",
    "    subprocess.run(cmd, shell=True)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "dd8baaf6-4074-499d-893c-3619c44f25a9",
   "metadata": {},
   "source": [
    "Construct rawdist trees"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "id": "a4ca0dff-5418-4e4a-8464-ea07c48b18ca",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>n_samples</th>\n",
       "      <th>genome_len</th>\n",
       "      <th>miss_threshold</th>\n",
       "      <th>pass_miss_threshold_sites</th>\n",
       "      <th>variant_sites</th>\n",
       "      <th>minimum_variant_strains_to_consider</th>\n",
       "      <th>considered_variant_sites</th>\n",
       "      <th>bed_mask_sites</th>\n",
       "      <th>passing_invariant_sites</th>\n",
       "      <th>passing_variant_sites</th>\n",
       "      <th>passing_output_sites</th>\n",
       "      <th>gb_memory</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>L1.1</th>\n",
       "      <td>2608.0</td>\n",
       "      <td>4411532.0</td>\n",
       "      <td>0.5</td>\n",
       "      <td>4318060.0</td>\n",
       "      <td>240888.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>240888.0</td>\n",
       "      <td>276750.0</td>\n",
       "      <td>3894806.0</td>\n",
       "      <td>227578.0</td>\n",
       "      <td>227578.0</td>\n",
       "      <td>21</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>L1.2</th>\n",
       "      <td>2034.0</td>\n",
       "      <td>4411532.0</td>\n",
       "      <td>0.5</td>\n",
       "      <td>4336432.0</td>\n",
       "      <td>163309.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>163309.0</td>\n",
       "      <td>276750.0</td>\n",
       "      <td>3976709.0</td>\n",
       "      <td>153275.0</td>\n",
       "      <td>153275.0</td>\n",
       "      <td>14</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>L2.1</th>\n",
       "      <td>390.0</td>\n",
       "      <td>4411532.0</td>\n",
       "      <td>0.5</td>\n",
       "      <td>4318472.0</td>\n",
       "      <td>64422.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>64422.0</td>\n",
       "      <td>276750.0</td>\n",
       "      <td>4069288.0</td>\n",
       "      <td>59716.0</td>\n",
       "      <td>59716.0</td>\n",
       "      <td>6</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>L2.2</th>\n",
       "      <td>16402.0</td>\n",
       "      <td>4411532.0</td>\n",
       "      <td>0.5</td>\n",
       "      <td>4313316.0</td>\n",
       "      <td>258399.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>258399.0</td>\n",
       "      <td>276750.0</td>\n",
       "      <td>3877983.0</td>\n",
       "      <td>243980.0</td>\n",
       "      <td>243980.0</td>\n",
       "      <td>111</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>L3</th>\n",
       "      <td>6313.0</td>\n",
       "      <td>4411532.0</td>\n",
       "      <td>0.5</td>\n",
       "      <td>4336143.0</td>\n",
       "      <td>306790.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>306790.0</td>\n",
       "      <td>276750.0</td>\n",
       "      <td>3841065.0</td>\n",
       "      <td>289219.0</td>\n",
       "      <td>289219.0</td>\n",
       "      <td>53</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>L4.1</th>\n",
       "      <td>9360.0</td>\n",
       "      <td>4411532.0</td>\n",
       "      <td>0.5</td>\n",
       "      <td>4323490.0</td>\n",
       "      <td>276656.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>276656.0</td>\n",
       "      <td>276750.0</td>\n",
       "      <td>3859384.0</td>\n",
       "      <td>260003.0</td>\n",
       "      <td>260003.0</td>\n",
       "      <td>69</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>L4.2</th>\n",
       "      <td>2783.0</td>\n",
       "      <td>4411532.0</td>\n",
       "      <td>0.5</td>\n",
       "      <td>4342269.0</td>\n",
       "      <td>104060.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>104060.0</td>\n",
       "      <td>276750.0</td>\n",
       "      <td>4037655.0</td>\n",
       "      <td>96993.0</td>\n",
       "      <td>96993.0</td>\n",
       "      <td>13</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>L4.3</th>\n",
       "      <td>8499.0</td>\n",
       "      <td>4411532.0</td>\n",
       "      <td>0.5</td>\n",
       "      <td>4324996.0</td>\n",
       "      <td>203742.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>203742.0</td>\n",
       "      <td>276750.0</td>\n",
       "      <td>3939635.0</td>\n",
       "      <td>191514.0</td>\n",
       "      <td>191514.0</td>\n",
       "      <td>48</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>L4.4</th>\n",
       "      <td>1869.0</td>\n",
       "      <td>4411532.0</td>\n",
       "      <td>0.5</td>\n",
       "      <td>4328080.0</td>\n",
       "      <td>104322.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>104322.0</td>\n",
       "      <td>276750.0</td>\n",
       "      <td>4028571.0</td>\n",
       "      <td>97138.0</td>\n",
       "      <td>97138.0</td>\n",
       "      <td>10</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>L4.5</th>\n",
       "      <td>872.0</td>\n",
       "      <td>4411532.0</td>\n",
       "      <td>0.5</td>\n",
       "      <td>4335074.0</td>\n",
       "      <td>98057.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>98057.0</td>\n",
       "      <td>276750.0</td>\n",
       "      <td>4036778.0</td>\n",
       "      <td>91279.0</td>\n",
       "      <td>91279.0</td>\n",
       "      <td>8</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>L4.6</th>\n",
       "      <td>1444.0</td>\n",
       "      <td>4411532.0</td>\n",
       "      <td>0.5</td>\n",
       "      <td>4330657.0</td>\n",
       "      <td>86741.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>86741.0</td>\n",
       "      <td>276750.0</td>\n",
       "      <td>4051656.0</td>\n",
       "      <td>80512.0</td>\n",
       "      <td>80512.0</td>\n",
       "      <td>9</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>L4.7</th>\n",
       "      <td>1040.0</td>\n",
       "      <td>4411532.0</td>\n",
       "      <td>0.5</td>\n",
       "      <td>4324710.0</td>\n",
       "      <td>83705.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>83705.0</td>\n",
       "      <td>276750.0</td>\n",
       "      <td>4050214.0</td>\n",
       "      <td>77773.0</td>\n",
       "      <td>77773.0</td>\n",
       "      <td>8</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>L4.8</th>\n",
       "      <td>4773.0</td>\n",
       "      <td>4411532.0</td>\n",
       "      <td>0.5</td>\n",
       "      <td>4332762.0</td>\n",
       "      <td>205909.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>205909.0</td>\n",
       "      <td>276750.0</td>\n",
       "      <td>3932673.0</td>\n",
       "      <td>192898.0</td>\n",
       "      <td>192898.0</td>\n",
       "      <td>30</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>L4.9</th>\n",
       "      <td>526.0</td>\n",
       "      <td>4411532.0</td>\n",
       "      <td>0.5</td>\n",
       "      <td>4338164.0</td>\n",
       "      <td>68933.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>68933.0</td>\n",
       "      <td>276750.0</td>\n",
       "      <td>4070584.0</td>\n",
       "      <td>64012.0</td>\n",
       "      <td>64012.0</td>\n",
       "      <td>6</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "      n_samples  genome_len  miss_threshold  pass_miss_threshold_sites  \\\n",
       "L1.1     2608.0   4411532.0             0.5                  4318060.0   \n",
       "L1.2     2034.0   4411532.0             0.5                  4336432.0   \n",
       "L2.1      390.0   4411532.0             0.5                  4318472.0   \n",
       "L2.2    16402.0   4411532.0             0.5                  4313316.0   \n",
       "L3       6313.0   4411532.0             0.5                  4336143.0   \n",
       "L4.1     9360.0   4411532.0             0.5                  4323490.0   \n",
       "L4.2     2783.0   4411532.0             0.5                  4342269.0   \n",
       "L4.3     8499.0   4411532.0             0.5                  4324996.0   \n",
       "L4.4     1869.0   4411532.0             0.5                  4328080.0   \n",
       "L4.5      872.0   4411532.0             0.5                  4335074.0   \n",
       "L4.6     1444.0   4411532.0             0.5                  4330657.0   \n",
       "L4.7     1040.0   4411532.0             0.5                  4324710.0   \n",
       "L4.8     4773.0   4411532.0             0.5                  4332762.0   \n",
       "L4.9      526.0   4411532.0             0.5                  4338164.0   \n",
       "\n",
       "      variant_sites  minimum_variant_strains_to_consider  \\\n",
       "L1.1       240888.0                                  1.0   \n",
       "L1.2       163309.0                                  1.0   \n",
       "L2.1        64422.0                                  1.0   \n",
       "L2.2       258399.0                                  1.0   \n",
       "L3         306790.0                                  1.0   \n",
       "L4.1       276656.0                                  1.0   \n",
       "L4.2       104060.0                                  1.0   \n",
       "L4.3       203742.0                                  1.0   \n",
       "L4.4       104322.0                                  1.0   \n",
       "L4.5        98057.0                                  1.0   \n",
       "L4.6        86741.0                                  1.0   \n",
       "L4.7        83705.0                                  1.0   \n",
       "L4.8       205909.0                                  1.0   \n",
       "L4.9        68933.0                                  1.0   \n",
       "\n",
       "      considered_variant_sites  bed_mask_sites  passing_invariant_sites  \\\n",
       "L1.1                  240888.0        276750.0                3894806.0   \n",
       "L1.2                  163309.0        276750.0                3976709.0   \n",
       "L2.1                   64422.0        276750.0                4069288.0   \n",
       "L2.2                  258399.0        276750.0                3877983.0   \n",
       "L3                    306790.0        276750.0                3841065.0   \n",
       "L4.1                  276656.0        276750.0                3859384.0   \n",
       "L4.2                  104060.0        276750.0                4037655.0   \n",
       "L4.3                  203742.0        276750.0                3939635.0   \n",
       "L4.4                  104322.0        276750.0                4028571.0   \n",
       "L4.5                   98057.0        276750.0                4036778.0   \n",
       "L4.6                   86741.0        276750.0                4051656.0   \n",
       "L4.7                   83705.0        276750.0                4050214.0   \n",
       "L4.8                  205909.0        276750.0                3932673.0   \n",
       "L4.9                   68933.0        276750.0                4070584.0   \n",
       "\n",
       "      passing_variant_sites  passing_output_sites  gb_memory  \n",
       "L1.1               227578.0              227578.0         21  \n",
       "L1.2               153275.0              153275.0         14  \n",
       "L2.1                59716.0               59716.0          6  \n",
       "L2.2               243980.0              243980.0        111  \n",
       "L3                 289219.0              289219.0         53  \n",
       "L4.1               260003.0              260003.0         69  \n",
       "L4.2                96993.0               96993.0         13  \n",
       "L4.3               191514.0              191514.0         48  \n",
       "L4.4                97138.0               97138.0         10  \n",
       "L4.5                91279.0               91279.0          8  \n",
       "L4.6                80512.0               80512.0          9  \n",
       "L4.7                77773.0               77773.0          8  \n",
       "L4.8               192898.0              192898.0         30  \n",
       "L4.9                64012.0               64012.0          6  "
      ]
     },
     "execution_count": 5,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# collect fasta generation stats\n",
    "fasta_stats = []\n",
    "for subtree_name in subtree_list:\n",
    "    fasta_stats.append(pd.read_csv(f'{work_dir}/{subtree_name}/{subtree_name}.results.csv', index_col=0).T)\n",
    "fasta_stats = pd.concat(fasta_stats, axis=0)\n",
    "# estimate memory use\n",
    "fasta_stats.loc[:, 'gb_memory'] = np.ceil(5 + 21 * fasta_stats.n_samples * fasta_stats.passing_output_sites / 1e+9 * 1.25).astype(int)\n",
    "fasta_stats"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 6,
   "id": "f3736f64-68e6-4691-95f7-9d6ad89d2666",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Submitted batch job 46567048\n",
      "Submitted batch job 46567049\n",
      "Submitted batch job 46567050\n",
      "Submitted batch job 46567051\n",
      "Submitted batch job 46567052\n",
      "Submitted batch job 46567053\n",
      "Submitted batch job 46567054\n",
      "Submitted batch job 46567055\n",
      "Submitted batch job 46567056\n",
      "Submitted batch job 46567057\n",
      "Submitted batch job 46567058\n",
      "Submitted batch job 46567059\n",
      "Submitted batch job 46567060\n",
      "Submitted batch job 46567061\n"
     ]
    }
   ],
   "source": [
    "threads = 50\n",
    "for subtree_name, subtree_data in fasta_stats.iterrows():\n",
    "    cmd = f'\\\n",
    "        sbatch -c {threads} -p sapphire -t {min(72, subtree_data.gb_memory.astype(int))}:00:00 --mem={subtree_data.gb_memory.astype(int)}G -o {work_dir}/{subtree_name}/tree-%j.out --wrap=\"\\\n",
    "        export OMP_NUM_THREADS={threads} && \\\n",
    "        FastTreeMP -rawdist -nt {work_dir}/{subtree_name}/{subtree_name}.fasta > {work_dir}/{subtree_name}/{subtree_name}.nwk\"'\n",
    "    subprocess.run(cmd, shell=True)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "6fea9412-c14d-4cc6-ac08-1fe5baeaeac2",
   "metadata": {
    "jp-MarkdownHeadingCollapsed": true
   },
   "source": [
    "## Normalize Subtrees"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "1adbed3b-a9a9-41bb-b57a-2b5bc567ddd7",
   "metadata": {},
   "source": [
    "Normalize, root and ladderize subtrees."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "id": "41a4ddd2-03a1-4d34-970e-4846904333fb",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>n_samples</th>\n",
       "      <th>genome_len</th>\n",
       "      <th>miss_threshold</th>\n",
       "      <th>pass_miss_threshold_sites</th>\n",
       "      <th>variant_sites</th>\n",
       "      <th>minimum_variant_strains_to_consider</th>\n",
       "      <th>considered_variant_sites</th>\n",
       "      <th>bed_mask_sites</th>\n",
       "      <th>passing_invariant_sites</th>\n",
       "      <th>passing_variant_sites</th>\n",
       "      <th>passing_output_sites</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>L1.1</th>\n",
       "      <td>2608.0</td>\n",
       "      <td>4411532.0</td>\n",
       "      <td>0.5</td>\n",
       "      <td>4318060.0</td>\n",
       "      <td>240888.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>240888.0</td>\n",
       "      <td>276750.0</td>\n",
       "      <td>3894806.0</td>\n",
       "      <td>227578.0</td>\n",
       "      <td>227578.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>L1.2</th>\n",
       "      <td>2034.0</td>\n",
       "      <td>4411532.0</td>\n",
       "      <td>0.5</td>\n",
       "      <td>4336432.0</td>\n",
       "      <td>163309.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>163309.0</td>\n",
       "      <td>276750.0</td>\n",
       "      <td>3976709.0</td>\n",
       "      <td>153275.0</td>\n",
       "      <td>153275.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>L2.1</th>\n",
       "      <td>390.0</td>\n",
       "      <td>4411532.0</td>\n",
       "      <td>0.5</td>\n",
       "      <td>4318472.0</td>\n",
       "      <td>64422.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>64422.0</td>\n",
       "      <td>276750.0</td>\n",
       "      <td>4069288.0</td>\n",
       "      <td>59716.0</td>\n",
       "      <td>59716.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>L2.2</th>\n",
       "      <td>16402.0</td>\n",
       "      <td>4411532.0</td>\n",
       "      <td>0.5</td>\n",
       "      <td>4313316.0</td>\n",
       "      <td>258399.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>258399.0</td>\n",
       "      <td>276750.0</td>\n",
       "      <td>3877983.0</td>\n",
       "      <td>243980.0</td>\n",
       "      <td>243980.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>L3</th>\n",
       "      <td>6313.0</td>\n",
       "      <td>4411532.0</td>\n",
       "      <td>0.5</td>\n",
       "      <td>4336143.0</td>\n",
       "      <td>306790.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>306790.0</td>\n",
       "      <td>276750.0</td>\n",
       "      <td>3841065.0</td>\n",
       "      <td>289219.0</td>\n",
       "      <td>289219.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>L4.1</th>\n",
       "      <td>9360.0</td>\n",
       "      <td>4411532.0</td>\n",
       "      <td>0.5</td>\n",
       "      <td>4323490.0</td>\n",
       "      <td>276656.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>276656.0</td>\n",
       "      <td>276750.0</td>\n",
       "      <td>3859384.0</td>\n",
       "      <td>260003.0</td>\n",
       "      <td>260003.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>L4.2</th>\n",
       "      <td>2783.0</td>\n",
       "      <td>4411532.0</td>\n",
       "      <td>0.5</td>\n",
       "      <td>4342269.0</td>\n",
       "      <td>104060.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>104060.0</td>\n",
       "      <td>276750.0</td>\n",
       "      <td>4037655.0</td>\n",
       "      <td>96993.0</td>\n",
       "      <td>96993.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>L4.3</th>\n",
       "      <td>8499.0</td>\n",
       "      <td>4411532.0</td>\n",
       "      <td>0.5</td>\n",
       "      <td>4324996.0</td>\n",
       "      <td>203742.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>203742.0</td>\n",
       "      <td>276750.0</td>\n",
       "      <td>3939635.0</td>\n",
       "      <td>191514.0</td>\n",
       "      <td>191514.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>L4.4</th>\n",
       "      <td>1869.0</td>\n",
       "      <td>4411532.0</td>\n",
       "      <td>0.5</td>\n",
       "      <td>4328080.0</td>\n",
       "      <td>104322.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>104322.0</td>\n",
       "      <td>276750.0</td>\n",
       "      <td>4028571.0</td>\n",
       "      <td>97138.0</td>\n",
       "      <td>97138.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>L4.5</th>\n",
       "      <td>872.0</td>\n",
       "      <td>4411532.0</td>\n",
       "      <td>0.5</td>\n",
       "      <td>4335074.0</td>\n",
       "      <td>98057.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>98057.0</td>\n",
       "      <td>276750.0</td>\n",
       "      <td>4036778.0</td>\n",
       "      <td>91279.0</td>\n",
       "      <td>91279.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>L4.6</th>\n",
       "      <td>1444.0</td>\n",
       "      <td>4411532.0</td>\n",
       "      <td>0.5</td>\n",
       "      <td>4330657.0</td>\n",
       "      <td>86741.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>86741.0</td>\n",
       "      <td>276750.0</td>\n",
       "      <td>4051656.0</td>\n",
       "      <td>80512.0</td>\n",
       "      <td>80512.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>L4.7</th>\n",
       "      <td>1040.0</td>\n",
       "      <td>4411532.0</td>\n",
       "      <td>0.5</td>\n",
       "      <td>4324710.0</td>\n",
       "      <td>83705.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>83705.0</td>\n",
       "      <td>276750.0</td>\n",
       "      <td>4050214.0</td>\n",
       "      <td>77773.0</td>\n",
       "      <td>77773.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>L4.8</th>\n",
       "      <td>4773.0</td>\n",
       "      <td>4411532.0</td>\n",
       "      <td>0.5</td>\n",
       "      <td>4332762.0</td>\n",
       "      <td>205909.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>205909.0</td>\n",
       "      <td>276750.0</td>\n",
       "      <td>3932673.0</td>\n",
       "      <td>192898.0</td>\n",
       "      <td>192898.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>L4.9</th>\n",
       "      <td>526.0</td>\n",
       "      <td>4411532.0</td>\n",
       "      <td>0.5</td>\n",
       "      <td>4338164.0</td>\n",
       "      <td>68933.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>68933.0</td>\n",
       "      <td>276750.0</td>\n",
       "      <td>4070584.0</td>\n",
       "      <td>64012.0</td>\n",
       "      <td>64012.0</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "      n_samples  genome_len  miss_threshold  pass_miss_threshold_sites  \\\n",
       "L1.1     2608.0   4411532.0             0.5                  4318060.0   \n",
       "L1.2     2034.0   4411532.0             0.5                  4336432.0   \n",
       "L2.1      390.0   4411532.0             0.5                  4318472.0   \n",
       "L2.2    16402.0   4411532.0             0.5                  4313316.0   \n",
       "L3       6313.0   4411532.0             0.5                  4336143.0   \n",
       "L4.1     9360.0   4411532.0             0.5                  4323490.0   \n",
       "L4.2     2783.0   4411532.0             0.5                  4342269.0   \n",
       "L4.3     8499.0   4411532.0             0.5                  4324996.0   \n",
       "L4.4     1869.0   4411532.0             0.5                  4328080.0   \n",
       "L4.5      872.0   4411532.0             0.5                  4335074.0   \n",
       "L4.6     1444.0   4411532.0             0.5                  4330657.0   \n",
       "L4.7     1040.0   4411532.0             0.5                  4324710.0   \n",
       "L4.8     4773.0   4411532.0             0.5                  4332762.0   \n",
       "L4.9      526.0   4411532.0             0.5                  4338164.0   \n",
       "\n",
       "      variant_sites  minimum_variant_strains_to_consider  \\\n",
       "L1.1       240888.0                                  1.0   \n",
       "L1.2       163309.0                                  1.0   \n",
       "L2.1        64422.0                                  1.0   \n",
       "L2.2       258399.0                                  1.0   \n",
       "L3         306790.0                                  1.0   \n",
       "L4.1       276656.0                                  1.0   \n",
       "L4.2       104060.0                                  1.0   \n",
       "L4.3       203742.0                                  1.0   \n",
       "L4.4       104322.0                                  1.0   \n",
       "L4.5        98057.0                                  1.0   \n",
       "L4.6        86741.0                                  1.0   \n",
       "L4.7        83705.0                                  1.0   \n",
       "L4.8       205909.0                                  1.0   \n",
       "L4.9        68933.0                                  1.0   \n",
       "\n",
       "      considered_variant_sites  bed_mask_sites  passing_invariant_sites  \\\n",
       "L1.1                  240888.0        276750.0                3894806.0   \n",
       "L1.2                  163309.0        276750.0                3976709.0   \n",
       "L2.1                   64422.0        276750.0                4069288.0   \n",
       "L2.2                  258399.0        276750.0                3877983.0   \n",
       "L3                    306790.0        276750.0                3841065.0   \n",
       "L4.1                  276656.0        276750.0                3859384.0   \n",
       "L4.2                  104060.0        276750.0                4037655.0   \n",
       "L4.3                  203742.0        276750.0                3939635.0   \n",
       "L4.4                  104322.0        276750.0                4028571.0   \n",
       "L4.5                   98057.0        276750.0                4036778.0   \n",
       "L4.6                   86741.0        276750.0                4051656.0   \n",
       "L4.7                   83705.0        276750.0                4050214.0   \n",
       "L4.8                  205909.0        276750.0                3932673.0   \n",
       "L4.9                   68933.0        276750.0                4070584.0   \n",
       "\n",
       "      passing_variant_sites  passing_output_sites  \n",
       "L1.1               227578.0              227578.0  \n",
       "L1.2               153275.0              153275.0  \n",
       "L2.1                59716.0               59716.0  \n",
       "L2.2               243980.0              243980.0  \n",
       "L3                 289219.0              289219.0  \n",
       "L4.1               260003.0              260003.0  \n",
       "L4.2                96993.0               96993.0  \n",
       "L4.3               191514.0              191514.0  \n",
       "L4.4                97138.0               97138.0  \n",
       "L4.5                91279.0               91279.0  \n",
       "L4.6                80512.0               80512.0  \n",
       "L4.7                77773.0               77773.0  \n",
       "L4.8               192898.0              192898.0  \n",
       "L4.9                64012.0               64012.0  "
      ]
     },
     "execution_count": 5,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "work_dir = 'scratch/supertree_construction/'\n",
    "# collect fasta generation stats\n",
    "fasta_stats = []\n",
    "for subtree_name in subtree_list:\n",
    "    fasta_stats.append(pd.read_csv(f'{work_dir}/{subtree_name}/{subtree_name}.results.csv', index_col=0).T)\n",
    "fasta_stats = pd.concat(fasta_stats, axis=0)\n",
    "fasta_stats"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 6,
   "id": "906af588-c077-4544-b390-011261e1a6dc",
   "metadata": {},
   "outputs": [],
   "source": [
    "for subtree_name, subtree_data in fasta_stats.iterrows():\n",
    "    # load the tree\n",
    "    tree = dendropy.Tree.get_from_path(\n",
    "        f'{work_dir}/{subtree_name}/{subtree_name}.nwk', 'newick')\n",
    "    # root at canettii\n",
    "    tree.reroot_at_node(\n",
    "        tree.find_node_with_taxon_label('canettii'))\n",
    "    tree.prune_taxa_with_labels('canettii')\n",
    "    # ladderize\n",
    "    tree.ladderize()\n",
    "    # rescale edges\n",
    "    for edge in tree.edges():\n",
    "        if edge.length != None:\n",
    "            edge.length = edge.length * (subtree_data.passing_output_sites / subtree_data.genome_len)\n",
    "    tree.write(\n",
    "        path=f'{work_dir}/{subtree_name}/{subtree_name}.norm.nwk',\n",
    "        schema='newick')\n",
    "    # finally, trim out background strains\n",
    "    tree.prune_taxa_with_labels(background_strains.strain.loc[background_strains.subtree != subtree_name].values)\n",
    "    tree.write(\n",
    "        path=f'{work_dir}/{subtree_name}/{subtree_name}.prune.nwk',\n",
    "        schema='newick')"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "04319c6a-ff92-4402-be75-93601b31b5c0",
   "metadata": {
    "jp-MarkdownHeadingCollapsed": true
   },
   "source": [
    "## Generate scaffold tree"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "63080b93-5ae2-4fdb-bd6c-772adea8a83c",
   "metadata": {},
   "source": [
    "Load in pared subtrees and pull strains"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "id": "6faaaabd-1535-4573-9a38-108a836e53aa",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>strain</th>\n",
       "      <th>subtree</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>SAMEA7525868</td>\n",
       "      <td>L1.1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>SAMEA7527532</td>\n",
       "      <td>L1.1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>SAMEA5607347</td>\n",
       "      <td>L1.1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>SAMN36619493</td>\n",
       "      <td>L1.1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>SAMEA7526267</td>\n",
       "      <td>L1.1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>20</th>\n",
       "      <td>SAMN13568144</td>\n",
       "      <td>L4.9</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>21</th>\n",
       "      <td>SAMEA112800064</td>\n",
       "      <td>L4.9</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>22</th>\n",
       "      <td>SAMN09090410</td>\n",
       "      <td>L4.9</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>23</th>\n",
       "      <td>SAMEA112799821</td>\n",
       "      <td>L4.9</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>24</th>\n",
       "      <td>SAMEA3715530</td>\n",
       "      <td>L4.9</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>514 rows × 2 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "            strain subtree\n",
       "0     SAMEA7525868    L1.1\n",
       "1     SAMEA7527532    L1.1\n",
       "2     SAMEA5607347    L1.1\n",
       "3     SAMN36619493    L1.1\n",
       "4     SAMEA7526267    L1.1\n",
       "..             ...     ...\n",
       "20    SAMN13568144    L4.9\n",
       "21  SAMEA112800064    L4.9\n",
       "22    SAMN09090410    L4.9\n",
       "23  SAMEA112799821    L4.9\n",
       "24    SAMEA3715530    L4.9\n",
       "\n",
       "[514 rows x 2 columns]"
      ]
     },
     "execution_count": 4,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "target_level = 6\n",
    "work_dir = 'scratch/supertree_construction/'\n",
    "\n",
    "def getLeafSampling(tree, target_level, seed=None):\n",
    "    # recursive function to get to level\n",
    "    def stepToLevel(target_node, target_level, current_level=0):\n",
    "        if current_level == target_level:\n",
    "            return [target_node]\n",
    "        elif target_node.is_leaf():\n",
    "            return [target_node]\n",
    "        else:\n",
    "            return np.concatenate(\n",
    "                [stepToLevel(child, target_level, current_level + 1)\n",
    "                 for child in target_node.child_nodes()])\n",
    "    np.random.seed(seed)\n",
    "    selected_leaves = []\n",
    "    for node in stepToLevel(tree.seed_node, target_level):\n",
    "        if node.is_leaf():\n",
    "            selected_leaves.append(node.taxon.label)\n",
    "        else:\n",
    "            selected_leaves.append(\n",
    "                np.random.choice(node.leaf_nodes()).taxon.label)\n",
    "    return selected_leaves\n",
    "\n",
    "# select supertree leaves\n",
    "seed = 42\n",
    "scaffold_data = []\n",
    "for subtree_name in subtree_list:\n",
    "    loaded_subtree = dendropy.Tree.get_from_path(\n",
    "        f'{work_dir}/{subtree_name}/{subtree_name}.prune.nwk','newick')\n",
    "    target_leaves = getLeafSampling(loaded_subtree, target_level, seed=seed)\n",
    "    scaffold_data.append(pd.DataFrame(\n",
    "        data=[target_leaves, [subtree_name for i in target_leaves]],\n",
    "        index=['strain', 'subtree']).T)\n",
    "scaffold_strains = pd.concat(scaffold_data, axis=0)\n",
    "scaffold_strains"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "99d00daf-974d-4c7f-8963-16a7867fd46a",
   "metadata": {},
   "source": [
    "Generate fasta"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 16,
   "id": "4bbabe93-ccd6-48f1-bc8f-234e81583f6d",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Submitted batch job 46868504\n"
     ]
    },
    {
     "data": {
      "text/plain": [
       "CompletedProcess(args='    sbatch -c 24 -p sapphire -t 1:00:00 --mem=50G -o scratch/supertree_construction//scaffold_1/fasta-%j.out --wrap=\"    module load Mambaforge/22.11.1-fasrc01 && conda activate mtb_isolates &&     export PATH=\"/n/boslfs02/LABS/sfortune_lab/Lab/culviner/bin/mtbvartools/scripts:/n/home12/pculviner/.conda/envs/mtb_isolates/bin/:$PATH\" &&     export PYTHONPATH=\"${PYTHONPATH}:/n/boslfs02/LABS/sfortune_lab/Lab/culviner/bin/mtbvartools\" &&     write_snp_fastas.py     --input-csv scratch/supertree_construction//scaffold_1/scaffold_inputs.csv     --input-fasta /n/boslfs02/LABS/sfortune_lab/Lab/culviner/notebooks/240807_isolate_tree_expansion/pipeline_inputs/Mtb_h37rv.fasta     --inputs-dir /n/boslfs02/LABS/sfortune_lab/Lab/culviner/notebooks/240807_isolate_tree_expansion/scratch/     --out-dir scratch/supertree_construction//scaffold_1/     --output scaffold     --mask rlc_plus_lowmap_marin.bed     --miss-threshold 0.5     --local-threads 10\"', returncode=0)"
      ]
     },
     "execution_count": 16,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# collect scaffold data\n",
    "scaffold_data = [[\n",
    "    'canettii',\n",
    "    '240805_qingyun_sra/SRR10522783/results/SRR10522783.breseq.vcf',\n",
    "    '240805_qingyun_sra/SRR10522783/results/SRR10522783.miss.breseq.zarr']]\n",
    "for index in np.unique(np.concatenate([scaffold_strains.strain.values, background_strains.strain.values])):\n",
    "    row_data = [index,\n",
    "         f'{subtree_inputs.loc[index, \"batch\"]}/{index}/results/{index}.breseq.vcf',\n",
    "         f'{subtree_inputs.loc[index, \"batch\"]}/{index}/results/{index}.miss.breseq.zarr']\n",
    "    scaffold_data.append(row_data)\n",
    "scaffold_df = pd.DataFrame(\n",
    "    data=scaffold_data,\n",
    "    columns=['label', 'vcf_path', 'miss_path'])\n",
    "\n",
    "# write df to working directory, but remove anything that's already there\n",
    "if os.path.exists(f'{work_dir}/scaffold_1'):\n",
    "    shutil.rmtree(f'{work_dir}/scaffold_1')\n",
    "os.makedirs(f'{work_dir}/scaffold_1', exist_ok=True)\n",
    "scaffold_df.to_csv(f'{work_dir}/scaffold_1/scaffold_inputs.csv')\n",
    "# run fasta generation command for subtree\n",
    "cmd = f'\\\n",
    "    sbatch -c 24 -p sapphire -t 1:00:00 --mem=50G -o {work_dir}/scaffold_1/fasta-%j.out --wrap=\"\\\n",
    "    module load Mambaforge/22.11.1-fasrc01 && conda activate mtb_isolates && \\\n",
    "    export PATH=\\\"/n/boslfs02/LABS/sfortune_lab/Lab/culviner/bin/mtbvartools/scripts:/n/home12/pculviner/.conda/envs/mtb_isolates/bin/:$PATH\\\" && \\\n",
    "    export PYTHONPATH=\\\"${{PYTHONPATH}}:/n/boslfs02/LABS/sfortune_lab/Lab/culviner/bin/mtbvartools\\\" && \\\n",
    "    write_snp_fastas.py \\\n",
    "    --input-csv {work_dir}/scaffold_1/scaffold_inputs.csv \\\n",
    "    --input-fasta /n/boslfs02/LABS/sfortune_lab/Lab/culviner/notebooks/240807_isolate_tree_expansion/pipeline_inputs/Mtb_h37rv.fasta \\\n",
    "    --inputs-dir /n/boslfs02/LABS/sfortune_lab/Lab/culviner/notebooks/240807_isolate_tree_expansion/scratch/ \\\n",
    "    --out-dir {work_dir}/scaffold_1/ \\\n",
    "    --output scaffold \\\n",
    "    --mask rlc_plus_lowmap_marin.bed \\\n",
    "    --miss-threshold 0.5 \\\n",
    "    --local-threads 10\"'\n",
    "subprocess.run(cmd, shell=True)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "id": "8c37b8b5-c3ef-4c24-a9d9-b4239a2a22b5",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "n_samples                                  785.0\n",
       "genome_len                             4411532.0\n",
       "miss_threshold                               0.5\n",
       "pass_miss_threshold_sites              4339859.0\n",
       "variant_sites                           129450.0\n",
       "minimum_variant_strains_to_consider          1.0\n",
       "considered_variant_sites                129450.0\n",
       "bed_mask_sites                          276750.0\n",
       "passing_invariant_sites                4013003.0\n",
       "passing_variant_sites                   121243.0\n",
       "passing_output_sites                    121243.0\n",
       "gb_memory                                    7.0\n",
       "Name: scaffold, dtype: float64"
      ]
     },
     "execution_count": 5,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "fasta_data = pd.read_csv(f'{work_dir}/scaffold_1/scaffold.results.csv', index_col=0).T.loc['scaffold']\n",
    "fasta_data.loc['gb_memory'] = np.ceil(5 + 21*fasta_data.n_samples * fasta_data.passing_output_sites / 1e+9).astype(int)\n",
    "fasta_data"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 19,
   "id": "f851a883-61ae-4e91-88bf-dcabbd9b94d2",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Submitted batch job 46868540\n"
     ]
    },
    {
     "data": {
      "text/plain": [
       "CompletedProcess(args='    sbatch -c 24 -p sapphire -t 1:00:00 --mem=7G -o scratch/supertree_construction//scaffold_1/tree-%j.out --wrap=\"    export OMP_NUM_THREADS=24 &&     FastTreeMP -rawdist -nt scratch/supertree_construction//scaffold_1/scaffold.fasta > scratch/supertree_construction//scaffold_1/scaffold.nwk\"', returncode=0)"
      ]
     },
     "execution_count": 19,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "threads = 24\n",
    "cmd = f'\\\n",
    "    sbatch -c {threads} -p sapphire -t 1:00:00 --mem={fasta_data.gb_memory.astype(int)}G -o {work_dir}/scaffold_1/tree-%j.out --wrap=\"\\\n",
    "    export OMP_NUM_THREADS={threads} && \\\n",
    "    FastTreeMP -rawdist -nt {work_dir}/scaffold_1/scaffold.fasta > {work_dir}/scaffold_1/scaffold.nwk\"'\n",
    "subprocess.run(cmd, shell=True)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 6,
   "id": "b7038939-396e-40fd-b08b-a31221c04c6f",
   "metadata": {},
   "outputs": [],
   "source": [
    "# load the tree\n",
    "tree = dendropy.Tree.get_from_path(\n",
    "    f'{work_dir}/scaffold_1/scaffold.nwk', 'newick')\n",
    "\n",
    "# root at canettii parent\n",
    "tree.reroot_at_node(\n",
    "    tree.find_node_with_taxon_label('canettii').parent_node)\n",
    "\n",
    "# ladderize\n",
    "tree.ladderize()\n",
    "# rescale edges\n",
    "for edge in tree.edges():\n",
    "    if edge.length != None:\n",
    "        edge.length = edge.length * (fasta_data.passing_output_sites / fasta_data.genome_len)\n",
    "tree.write(\n",
    "    path=f'{work_dir}/scaffold_1/scaffold.outgroup.nwk',\n",
    "    schema='newick')"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "c765d30c-eae7-40c5-905d-064ffa3f7618",
   "metadata": {},
   "source": [
    "Check for incorrect supertree assignments."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 7,
   "id": "b67bee9a-4802-4bd8-9bad-d53fd7e34167",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "L1.1: ['L1.1']\n",
      "L1.2: ['L1.2']\n",
      "L2.1: ['L2.1']\n",
      "L2.2: ['L2.2']\n",
      "L3: ['L3']\n",
      "L4.1: ['L4.1']\n",
      "L4.2: ['L4.2']\n",
      "L4.3: ['L4.3']\n",
      "L4.4: ['L4.4']\n",
      "L4.5: ['L4.5']\n",
      "L4.6: ['L4.6']\n",
      "L4.7: ['L4.7']\n",
      "L4.8: ['L4.8']\n",
      "L4.9: ['L4.9']\n"
     ]
    }
   ],
   "source": [
    "for subtree in subtree_list:\n",
    "    mrca_node = tree.mrca(taxon_labels=scaffold_strains.strain[scaffold_strains.subtree == subtree].values)\n",
    "    mrca_leaves = [n.taxon.label for n in mrca_node.leaf_nodes()]\n",
    "    subtree_mrca_child_nodes = np.unique(np.concatenate([\n",
    "        scaffold_strains.subtree[np.isin(scaffold_strains.strain, mrca_leaves)].values,\n",
    "        background_strains.subtree[np.isin(background_strains.strain, mrca_leaves)].values]))\n",
    "    print(f'{subtree}: {subtree_mrca_child_nodes}')"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "1d6544a5-27f0-4023-b18a-e29d850ded2f",
   "metadata": {},
   "source": [
    "Paste together supertree from scaffold (pruned)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 8,
   "id": "14e178f9-963d-4570-801f-0be658aed9d5",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "14 14\n"
     ]
    }
   ],
   "source": [
    "# first remove scaffold branches\n",
    "scaffold_tree = dendropy.Tree.get_from_path(\n",
    "    f'{work_dir}/scaffold_1/scaffold.outgroup.nwk', 'newick')\n",
    "node_lookups = {}\n",
    "for subtree_name in subtree_list:\n",
    "    mrca_node = scaffold_tree.mrca(taxon_labels=scaffold_strains.strain[scaffold_strains.subtree == subtree_name].values)\n",
    "    # remove mrca child nodes\n",
    "    for child_node in mrca_node.child_nodes():\n",
    "        mrca_node.remove_child(child_node)\n",
    "    node_lookups[subtree_name] = mrca_node\n",
    "\n",
    "# verify number of nodes\n",
    "print(len(node_lookups), len(subtree_list))\n",
    "\n",
    "# now paste on subtrees to scaffold\n",
    "for subtree_name in subtree_list:\n",
    "    mrca_node = node_lookups[subtree_name]\n",
    "    # load subtree\n",
    "    loaded_subtree = dendropy.Tree.get_from_path(\n",
    "        f'{work_dir}/{subtree_name}/{subtree_name}.prune.nwk', 'newick')\n",
    "    for child_node in loaded_subtree.seed_node.child_nodes():\n",
    "        mrca_node.add_child(child_node)\n",
    "\n",
    "scaffold_tree.ladderize()\n",
    "scaffold_tree.write(\n",
    "    path=f'{work_dir}/scaffold_1/reconstructed_supertree.nwk',\n",
    "    schema='newick')"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "b247523a-7387-4d47-8dbc-6c1d1abffde2",
   "metadata": {},
   "source": [
    "# Tree Generation Testing"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "7c362a33-20e8-49cf-91ce-5d5d78881c6f",
   "metadata": {
    "jp-MarkdownHeadingCollapsed": true
   },
   "source": [
    "## Generate scaffold tree (2)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "bf75e623-8624-48ea-ae24-931d660147a8",
   "metadata": {},
   "source": [
    "Load in pared subtrees and pull strains"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 8,
   "id": "d8121d79-2d99-4648-b3f2-897ab55194cd",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>strain</th>\n",
       "      <th>subtree</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>SAMEA7525868</td>\n",
       "      <td>L1.1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>SAMEA7528466</td>\n",
       "      <td>L1.1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>SAMEA11225748</td>\n",
       "      <td>L1.1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>SAMN36619493</td>\n",
       "      <td>L1.1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>SAMEA11225683</td>\n",
       "      <td>L1.1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>20</th>\n",
       "      <td>SAMN13568144</td>\n",
       "      <td>L4.9</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>21</th>\n",
       "      <td>SAMEA112800064</td>\n",
       "      <td>L4.9</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>22</th>\n",
       "      <td>SAMEA1101738</td>\n",
       "      <td>L4.9</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>23</th>\n",
       "      <td>SAMEA112799821</td>\n",
       "      <td>L4.9</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>24</th>\n",
       "      <td>SAMEA7529273</td>\n",
       "      <td>L4.9</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>514 rows × 2 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "            strain subtree\n",
       "0     SAMEA7525868    L1.1\n",
       "1     SAMEA7528466    L1.1\n",
       "2    SAMEA11225748    L1.1\n",
       "3     SAMN36619493    L1.1\n",
       "4    SAMEA11225683    L1.1\n",
       "..             ...     ...\n",
       "20    SAMN13568144    L4.9\n",
       "21  SAMEA112800064    L4.9\n",
       "22    SAMEA1101738    L4.9\n",
       "23  SAMEA112799821    L4.9\n",
       "24    SAMEA7529273    L4.9\n",
       "\n",
       "[514 rows x 2 columns]"
      ]
     },
     "execution_count": 8,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "target_level = 6\n",
    "work_dir = 'scratch/supertree_construction/'\n",
    "\n",
    "def getLeafSampling(tree, target_level, seed=None):\n",
    "    # recursive function to get to level\n",
    "    def stepToLevel(target_node, target_level, current_level=0):\n",
    "        if current_level == target_level:\n",
    "            return [target_node]\n",
    "        elif target_node.is_leaf():\n",
    "            return [target_node]\n",
    "        else:\n",
    "            return np.concatenate(\n",
    "                [stepToLevel(child, target_level, current_level + 1)\n",
    "                 for child in target_node.child_nodes()])\n",
    "    np.random.seed(seed)\n",
    "    selected_leaves = []\n",
    "    for node in stepToLevel(tree.seed_node, target_level):\n",
    "        if node.is_leaf():\n",
    "            selected_leaves.append(node.taxon.label)\n",
    "        else:\n",
    "            selected_leaves.append(\n",
    "                np.random.choice(node.leaf_nodes()).taxon.label)\n",
    "    return selected_leaves\n",
    "\n",
    "# select supertree leaves\n",
    "seed = 42\n",
    "scaffold_data = []\n",
    "for subtree_name in subtree_list:\n",
    "    loaded_subtree = dendropy.Tree.get_from_path(\n",
    "        f'{work_dir}/{subtree_name}/{subtree_name}.prune.nwk','newick')\n",
    "    target_leaves = getLeafSampling(loaded_subtree, target_level, seed=seed)\n",
    "    scaffold_data.append(pd.DataFrame(\n",
    "        data=[target_leaves, [subtree_name for i in target_leaves]],\n",
    "        index=['strain', 'subtree']).T)\n",
    "scaffold_strains = pd.concat(scaffold_data, axis=0)\n",
    "scaffold_strains"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "49a16339-21b9-47df-b2b6-ccbabb407592",
   "metadata": {},
   "source": [
    "Generate fasta"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 10,
   "id": "86e42361-8af0-479d-acb0-efa56efdc0a5",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Submitted batch job 46867813\n"
     ]
    },
    {
     "data": {
      "text/plain": [
       "CompletedProcess(args='    sbatch -c 24 -p sapphire -t 1:00:00 --mem=50G -o scratch/supertree_construction//scaffold_1/fasta-%j.out --wrap=\"    module load Mambaforge/22.11.1-fasrc01 && conda activate mtb_isolates &&     export PATH=\"/n/boslfs02/LABS/sfortune_lab/Lab/culviner/bin/mtbvartools/scripts:/n/home12/pculviner/.conda/envs/mtb_isolates/bin/:$PATH\" &&     export PYTHONPATH=\"${PYTHONPATH}:/n/boslfs02/LABS/sfortune_lab/Lab/culviner/bin/mtbvartools\" &&     write_snp_fastas.py     --input-csv scratch/supertree_construction//scaffold_1/scaffold_inputs.csv     --input-fasta /n/boslfs02/LABS/sfortune_lab/Lab/culviner/notebooks/240807_isolate_tree_expansion/pipeline_inputs/Mtb_h37rv.fasta     --inputs-dir /n/boslfs02/LABS/sfortune_lab/Lab/culviner/notebooks/240807_isolate_tree_expansion/scratch/     --out-dir scratch/supertree_construction//scaffold_1/     --output scaffold     --mask rlc_plus_lowmap_marin.bed     --miss-threshold 0.5     --local-threads 10\"', returncode=0)"
      ]
     },
     "execution_count": 10,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# collect scaffold data\n",
    "scaffold_data = [[\n",
    "    'canettii',\n",
    "    '240805_qingyun_sra/SRR10522783/results/SRR10522783.breseq.vcf',\n",
    "    '240805_qingyun_sra/SRR10522783/results/SRR10522783.miss.breseq.zarr']]\n",
    "for index in np.unique(np.concatenate([scaffold_strains.strain.values, background_strains.strain.values])):\n",
    "    row_data = [index,\n",
    "         f'{subtree_inputs.loc[index, \"batch\"]}/{index}/results/{index}.breseq.vcf',\n",
    "         f'{subtree_inputs.loc[index, \"batch\"]}/{index}/results/{index}.miss.breseq.zarr']\n",
    "    scaffold_data.append(row_data)\n",
    "scaffold_df = pd.DataFrame(\n",
    "    data=scaffold_data,\n",
    "    columns=['label', 'vcf_path', 'miss_path'])\n",
    "\n",
    "# write df to working directory, but remove anything that's already there\n",
    "if os.path.exists(f'{work_dir}/scaffold_1'):\n",
    "    shutil.rmtree(f'{work_dir}/scaffold_1')\n",
    "os.makedirs(f'{work_dir}/scaffold_1', exist_ok=True)\n",
    "scaffold_df.to_csv(f'{work_dir}/scaffold_1/scaffold_inputs.csv')\n",
    "# run fasta generation command for subtree\n",
    "cmd = f'\\\n",
    "    sbatch -c 24 -p sapphire -t 1:00:00 --mem=50G -o {work_dir}/scaffold_1/fasta-%j.out --wrap=\"\\\n",
    "    module load Mambaforge/22.11.1-fasrc01 && conda activate mtb_isolates && \\\n",
    "    export PATH=\\\"/n/boslfs02/LABS/sfortune_lab/Lab/culviner/bin/mtbvartools/scripts:/n/home12/pculviner/.conda/envs/mtb_isolates/bin/:$PATH\\\" && \\\n",
    "    export PYTHONPATH=\\\"${{PYTHONPATH}}:/n/boslfs02/LABS/sfortune_lab/Lab/culviner/bin/mtbvartools\\\" && \\\n",
    "    write_snp_fastas.py \\\n",
    "    --input-csv {work_dir}/scaffold_1/scaffold_inputs.csv \\\n",
    "    --input-fasta /n/boslfs02/LABS/sfortune_lab/Lab/culviner/notebooks/240807_isolate_tree_expansion/pipeline_inputs/Mtb_h37rv.fasta \\\n",
    "    --inputs-dir /n/boslfs02/LABS/sfortune_lab/Lab/culviner/notebooks/240807_isolate_tree_expansion/scratch/ \\\n",
    "    --out-dir {work_dir}/scaffold_1/ \\\n",
    "    --output scaffold \\\n",
    "    --mask rlc_plus_lowmap_marin.bed \\\n",
    "    --miss-threshold 0.5 \\\n",
    "    --local-threads 10\"'\n",
    "subprocess.run(cmd, shell=True)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 11,
   "id": "93efa942-3732-4d96-8c90-49eb8abee09b",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "n_samples                                  787.0\n",
       "genome_len                             4411532.0\n",
       "miss_threshold                               0.5\n",
       "pass_miss_threshold_sites              4339863.0\n",
       "variant_sites                           129150.0\n",
       "minimum_variant_strains_to_consider          1.0\n",
       "considered_variant_sites                129150.0\n",
       "bed_mask_sites                          276750.0\n",
       "passing_invariant_sites                4013258.0\n",
       "passing_variant_sites                   120952.0\n",
       "passing_output_sites                    120952.0\n",
       "gb_memory                                    7.0\n",
       "Name: scaffold, dtype: float64"
      ]
     },
     "execution_count": 11,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "fasta_data = pd.read_csv(f'{work_dir}/scaffold_1/scaffold.results.csv', index_col=0).T.loc['scaffold']\n",
    "fasta_data.loc['gb_memory'] = np.ceil(5 + 21*fasta_data.n_samples * fasta_data.passing_output_sites / 1e+9).astype(int)\n",
    "fasta_data"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 12,
   "id": "7c845f02-ca7d-41f9-bf1e-c882d37fb6d4",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Submitted batch job 46867848\n"
     ]
    },
    {
     "data": {
      "text/plain": [
       "CompletedProcess(args='    sbatch -c 24 -p sapphire -t 1:00:00 --mem=7G -o scratch/supertree_construction//scaffold_1/tree-%j.out --wrap=\"    export OMP_NUM_THREADS=24 &&     FastTreeMP -rawdist -nt scratch/supertree_construction//scaffold_1/scaffold.fasta > scratch/supertree_construction//scaffold_1/scaffold.nwk\"', returncode=0)"
      ]
     },
     "execution_count": 12,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "threads = 24\n",
    "cmd = f'\\\n",
    "    sbatch -c {threads} -p sapphire -t 1:00:00 --mem={fasta_data.gb_memory.astype(int)}G -o {work_dir}/scaffold_1/tree-%j.out --wrap=\"\\\n",
    "    export OMP_NUM_THREADS={threads} && \\\n",
    "    FastTreeMP -rawdist -nt {work_dir}/scaffold_1/scaffold.fasta > {work_dir}/scaffold_1/scaffold.nwk\"'\n",
    "subprocess.run(cmd, shell=True)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "64fdcd90-dc1f-4e0d-9220-23b6e7528c41",
   "metadata": {},
   "source": [
    "Normalize the tree."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 7,
   "id": "0f7b9c20-7367-42a9-be71-f0e6bd958999",
   "metadata": {},
   "outputs": [],
   "source": [
    "# load the tree\n",
    "tree = dendropy.Tree.get_from_path(\n",
    "    f'{work_dir}/scaffold_1/scaffold.nwk', 'newick')\n",
    "# root at canettii\n",
    "tree.reroot_at_node(\n",
    "    tree.find_node_with_taxon_label('canettii'))\n",
    "tree.prune_taxa_with_labels('canettii')\n",
    "# ladderize\n",
    "tree.ladderize()\n",
    "# rescale edges\n",
    "for edge in tree.edges():\n",
    "    if edge.length != None:\n",
    "        edge.length = edge.length * (fasta_data.passing_output_sites / fasta_data.genome_len)\n",
    "tree.write(\n",
    "    path=f'{work_dir}/scaffold_1/scaffold.norm.nwk',\n",
    "    schema='newick')"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "f1906573-28cd-42c7-9a50-7fb3d7970620",
   "metadata": {},
   "source": [
    "### Generate a complete supertree"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 65,
   "id": "fb183756-40d6-49ad-87aa-1033f0be2c48",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Submitted batch job 46448454\n"
     ]
    },
    {
     "data": {
      "text/plain": [
       "CompletedProcess(args='    sbatch -c 10 -p shared -t 1:00:00 --mem=20G -o scratch/subtree_test//L4.9/fasta-%j.out --wrap=\"    module load Mambaforge/22.11.1-fasrc01 && conda activate mtb_isolates &&     export PATH=\"/n/boslfs02/LABS/sfortune_lab/Lab/culviner/bin/mtbvartools/scripts:/n/home12/pculviner/.conda/envs/mtb_isolates/bin/:$PATH\" &&     export PYTHONPATH=\"${PYTHONPATH}:/n/boslfs02/LABS/sfortune_lab/Lab/culviner/bin/mtbvartools\" &&     write_snp_fastas.py     --input-csv scratch/subtree_test//supertree/supertree_inputs.csv     --input-fasta /n/boslfs02/LABS/sfortune_lab/Lab/culviner/notebooks/240807_isolate_tree_expansion/pipeline_inputs/Mtb_h37rv.fasta     --inputs-dir /n/boslfs02/LABS/sfortune_lab/Lab/culviner/notebooks/240807_isolate_tree_expansion/scratch/     --out-dir scratch/subtree_test//supertree/     --output supertree     --mask rlc_plus_lowmap_marin.bed     --miss-threshold 0.5     --local-threads 10\"', returncode=0)"
      ]
     },
     "execution_count": 65,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "supertree_data = [[\n",
    "    'canettii',\n",
    "    '240805_qingyun_sra/SRR10522783/results/SRR10522783.breseq.vcf',\n",
    "    '240805_qingyun_sra/SRR10522783/results/SRR10522783.miss.breseq.zarr']]\n",
    "for index in all_leaves:\n",
    "    row_data = [index,\n",
    "         f'{subtree_inputs.loc[index, \"batch\"]}/{index}/results/{index}.breseq.vcf',\n",
    "         f'{subtree_inputs.loc[index, \"batch\"]}/{index}/results/{index}.miss.breseq.zarr']\n",
    "    supertree_data.append(row_data)\n",
    "supertree_df = pd.DataFrame(\n",
    "    data=supertree_data,\n",
    "    columns=['label', 'vcf_path', 'miss_path'])\n",
    "# write df to working directory, but remove anything that's already there\n",
    "if os.path.exists(f'{work_dir}/supertree'):\n",
    "    shutil.rmtree(f'{work_dir}/supertree')\n",
    "os.makedirs(f'{work_dir}/supertree', exist_ok=True)\n",
    "supertree_df.to_csv(f'{work_dir}/supertree/supertree_inputs.csv')\n",
    "# run fasta generation command for subtree\n",
    "cmd = f'\\\n",
    "    sbatch -c 10 -p shared -t 1:00:00 --mem=20G -o {work_dir}/supertree/fasta-%j.out --wrap=\"\\\n",
    "    module load Mambaforge/22.11.1-fasrc01 && conda activate mtb_isolates && \\\n",
    "    export PATH=\\\"/n/boslfs02/LABS/sfortune_lab/Lab/culviner/bin/mtbvartools/scripts:/n/home12/pculviner/.conda/envs/mtb_isolates/bin/:$PATH\\\" && \\\n",
    "    export PYTHONPATH=\\\"${{PYTHONPATH}}:/n/boslfs02/LABS/sfortune_lab/Lab/culviner/bin/mtbvartools\\\" && \\\n",
    "    write_snp_fastas.py \\\n",
    "    --input-csv {work_dir}/supertree/supertree_inputs.csv \\\n",
    "    --input-fasta /n/boslfs02/LABS/sfortune_lab/Lab/culviner/notebooks/240807_isolate_tree_expansion/pipeline_inputs/Mtb_h37rv.fasta \\\n",
    "    --inputs-dir /n/boslfs02/LABS/sfortune_lab/Lab/culviner/notebooks/240807_isolate_tree_expansion/scratch/ \\\n",
    "    --out-dir {work_dir}/supertree/ \\\n",
    "    --output supertree \\\n",
    "    --mask rlc_plus_lowmap_marin.bed \\\n",
    "    --miss-threshold 0.5 \\\n",
    "    --local-threads 10\"'\n",
    "subprocess.run(cmd, shell=True)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 69,
   "id": "9859cdb3-796d-4c0f-b4d6-2ffd46d19de9",
   "metadata": {},
   "outputs": [],
   "source": [
    "fasta_data = pd.read_csv(f'{work_dir}/supertree/supertree.results.csv', index_col=0).T.loc['supertree']\n",
    "fasta_data.loc['gb_memory'] = np.ceil(21*fasta_data.n_samples * fasta_data.passing_output_sites / 1e+9).astype(int)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 71,
   "id": "4f046b85-7743-475e-93b3-c7d637aec2b1",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Submitted batch job 46449354\n"
     ]
    },
    {
     "data": {
      "text/plain": [
       "CompletedProcess(args='    sbatch -c 24 -p shared -t 5:00:00 --mem=5G -o scratch/subtree_test//supertree/tree-%j.out --wrap=\"    export OMP_NUM_THREADS=24 &&     FastTreeMP -rawdist -nt scratch/subtree_test//supertree/supertree.fasta > scratch/subtree_test//supertree/supertree.nwk\"', returncode=0)"
      ]
     },
     "execution_count": 71,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "threads = 24\n",
    "cmd = f'\\\n",
    "    sbatch -c {threads} -p shared -t 5:00:00 --mem={fasta_data.gb_memory.astype(int)}G -o {work_dir}/supertree/tree-%j.out --wrap=\"\\\n",
    "    export OMP_NUM_THREADS={threads} && \\\n",
    "    FastTreeMP -rawdist -nt {work_dir}/supertree/supertree.fasta > {work_dir}/supertree/supertree.nwk\"'\n",
    "subprocess.run(cmd, shell=True)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 100,
   "id": "a90c5617-9f92-4533-b10a-e64c7ad02db6",
   "metadata": {},
   "outputs": [],
   "source": [
    "# load the tree\n",
    "tree = dendropy.Tree.get_from_path(\n",
    "    f'{work_dir}/supertree/supertree.nwk', 'newick')\n",
    "# root at canettii\n",
    "tree.reroot_at_node(\n",
    "    tree.find_node_with_taxon_label('canettii'))\n",
    "tree.prune_taxa_with_labels('canettii')\n",
    "# ladderize\n",
    "tree.ladderize()\n",
    "# rescale edges\n",
    "for edge in tree.edges():\n",
    "    if edge.length != None:\n",
    "        edge.length = edge.length * (fasta_data.passing_output_sites / fasta_data.genome_len)\n",
    "tree.write(\n",
    "    path=f'{work_dir}/supertree/supertree.norm.nwk',\n",
    "    schema='newick')\n",
    "all_leaves += [l.taxon.label for l in tree.leaf_nodes()]"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "e15bb8c8-cd15-4039-8dca-2735268eb9d7",
   "metadata": {},
   "source": [
    "### Compare tree branch lengths"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 101,
   "id": "bfd573d2-28a8-4d1b-8063-7cdce02a3cfb",
   "metadata": {},
   "outputs": [],
   "source": [
    "supertree = dendropy.Tree.get_from_path(\n",
    "    f'{work_dir}/supertree/supertree.norm.nwk', 'newick')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 102,
   "id": "ce589d99-80be-4fd5-9edc-fd99116cc838",
   "metadata": {},
   "outputs": [],
   "source": [
    "edge_comparisons = []\n",
    "for st_name in subtree_list:\n",
    "    subtree = dendropy.Tree.get_from_path(\n",
    "        f'{work_dir}/{st_name}/{st_name}.norm.nwk', 'newick')\n",
    "    for n in subtree.leaf_nodes():\n",
    "        supertree_len = supertree.find_node_with_taxon_label(n.taxon.label).edge_length\n",
    "        subtree_len = n.edge_length\n",
    "        edge_comparisons.append([supertree_len, subtree_len])\n",
    "\n",
    "edge_comparisons = pd.DataFrame(\n",
    "    data=edge_comparisons,\n",
    "    columns=['super_len', 'sub_len'])"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 106,
   "id": "7d1cdc29-51a5-40bc-8bf4-2519fc39ddcd",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>super_len</th>\n",
       "      <th>sub_len</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>5.485550e-05</td>\n",
       "      <td>5.485035e-05</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>2.641357e-05</td>\n",
       "      <td>2.640955e-05</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>3.066918e-05</td>\n",
       "      <td>3.088379e-05</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>5.216077e-05</td>\n",
       "      <td>5.215450e-05</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>6.422262e-05</td>\n",
       "      <td>6.432705e-05</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1395</th>\n",
       "      <td>0.000000e+00</td>\n",
       "      <td>0.000000e+00</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1396</th>\n",
       "      <td>1.551332e-10</td>\n",
       "      <td>2.821129e-11</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1397</th>\n",
       "      <td>1.134706e-06</td>\n",
       "      <td>1.134455e-06</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1398</th>\n",
       "      <td>1.551332e-10</td>\n",
       "      <td>2.821129e-11</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1399</th>\n",
       "      <td>1.861598e-10</td>\n",
       "      <td>2.821129e-11</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>1400 rows × 2 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "         super_len       sub_len\n",
       "0     5.485550e-05  5.485035e-05\n",
       "1     2.641357e-05  2.640955e-05\n",
       "2     3.066918e-05  3.088379e-05\n",
       "3     5.216077e-05  5.215450e-05\n",
       "4     6.422262e-05  6.432705e-05\n",
       "...            ...           ...\n",
       "1395  0.000000e+00  0.000000e+00\n",
       "1396  1.551332e-10  2.821129e-11\n",
       "1397  1.134706e-06  1.134455e-06\n",
       "1398  1.551332e-10  2.821129e-11\n",
       "1399  1.861598e-10  2.821129e-11\n",
       "\n",
       "[1400 rows x 2 columns]"
      ]
     },
     "execution_count": 106,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "edge_comparisons"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "bf5e8bd8-7956-4225-a657-34f61aa54d83",
   "metadata": {},
   "source": [
    "### Edit supertree to use subtree as branches"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 127,
   "id": "70415a31-c5c3-4d60-9258-05d394a29d25",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "<Tree object at 0x14f23a7125d0>"
      ]
     },
     "execution_count": 127,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "supertree = dendropy.Tree.get_from_path(\n",
    "    f'{work_dir}/supertree/supertree.norm.nwk', 'newick')\n",
    "supertree"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 146,
   "id": "09cb47d3-954a-4a12-84f3-37c4cbc14e20",
   "metadata": {},
   "outputs": [],
   "source": [
    "for st_name in subtree_list:\n",
    "    subtree = dendropy.Tree.get_from_path(  # get the subtree\n",
    "        f'{work_dir}/{st_name}/{st_name}.norm.nwk', 'newick')\n",
    "    mrca = supertree.mrca(  # get the supertree defining nodes on the subtree\n",
    "        taxon_labels=[l.taxon.label for l in subtree.leaf_nodes()])\n",
    "    supertree.prune_nodes(mrca.child_nodes())\n",
    "    supertree"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 148,
   "id": "2f57fce1-3b5e-43a5-a848-9ee147c47db3",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "[&R] ((:8.657455476918221e-06,:1.5513318275827988e-10)1.000:0.00011046742293833524,((:5.892035847750849e-05,(:5.193772084164866e-05,:3.073331072969662e-05)1.000:2.7092365956996346e-05)1.000:1.2485925240936708e-05,(:2.0194865411834256e-05,(:3.563195124165483e-05,(:2.66109256376243e-05,(:1.1800205546508559e-05,(:3.10831670834531e-05,(:3.176413970248884e-06,(:2.497613215771755e-06,(:1.633266969388412e-05,:4.766746279977115e-06)1.000:4.348600299170446e-06)1.000:2.0142368342788855e-05)1.000:2.716382030097481e-06)1.000:2.950478002879725e-06)0.997:1.5513318275827988e-10)0.424:1.3157217549368337e-05)1.000:2.1392865902366797e-07)0.740:3.701335018084421e-05)1.000:6.223754029779225e-05)canettii;\n"
     ]
    }
   ],
   "source": [
    "print(supertree)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "62c12f28-6995-4190-925e-e076a8626840",
   "metadata": {},
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "general_py",
   "language": "python",
   "name": "general_py"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.12.8"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 5
}
